beacon-chain/monitor/service.go

```diff
@@ -1,25 +1,55 @@
 package monitor
 
 import (
+	"context"
+
 	types "github.com/prysmaticlabs/eth2-types"
+	"github.com/prysmaticlabs/prysm/beacon-chain/state/stategen"
 	"github.com/sirupsen/logrus"
 )
 
 var (
 	log = logrus.WithField("prefix", "monitor")
 )
 
+// ValidatorLatestPerformance keeps track of the latest participation of the validator
+type ValidatorLatestPerformance struct {
+	attestedSlot  types.Slot
+	inclusionSlot types.Slot
+	timelySource  bool
+	timelyTarget  bool
+	timelyHead    bool
+	balance       uint64
+	balanceChange uint64
+}
+
+// ValidatorAggregatedPerformance keeps track of the accumulated performance of
+// the validator since launch
+type ValidatorAggregatedPerformance struct {
+	totalAttestedCount  uint64
+	totalRequestedCount uint64
+	totalDistance       uint64
+	totalCorrectSource  uint64
+	totalCorrectTarget  uint64
+	totalCorrectHead    uint64
+	totalAggregations   uint64
+}
+
 // ValidatorMonitorConfig contains the list of validator indices that the
 // monitor service tracks, as well as the event feed notifier that the
 // monitor needs to subscribe.
 type ValidatorMonitorConfig struct {
+	StateGen          stategen.StateManager
 	TrackedValidators map[types.ValidatorIndex]interface{}
 }
 
 // Service is the main structure that tracks validators and reports logs and
 // metrics of their performances throughout their lifetime.
 type Service struct {
-	config *ValidatorMonitorConfig
+	config                *ValidatorMonitorConfig
+	ctx                   context.Context
```

Comment:
How/when is this going to be used? In general, a global ctx is used to signal the shutdown of the service, especially when the service has some infinite interval loop.

Reply:
This will be the last PR, when we integrate the service.